### PR TITLE
Create fake /etc/hosts for conformance test

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

    "KubeletManagedEtcHosts should test kubelet managed /etc/hosts file"
    conformance test fails in the CI's Docker-In-Docker environment.

    This test mounts a /etc/hosts file and checks if "# Kubernetes-managed
    hosts file." string is present or not under various conditions. The
    specific failure with DIND happens when the /etc/hosts picked up
    from the box where e2e test are running already has this string. This
    happens because our CI runs on kubernetes and the e2e tests are running
    in a container that was started on kubernetes (and hence already has
    that string)

    To avoid this situation, we create a new /etc/hosts file with known
    contents (and does not have the "# Kubernetes-managed hosts file."
    string)

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Please see:
https://k8s-testgrid.appspot.com/sig-testing-misc#ci-kubernetes-local-e2e
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-local-e2e/285

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/common/kubelet_etc_hosts.go [hunk 1]
```diff
@@ -137,6 +139,27 @@ func (config *KubeletManagedHostConfig) getEtcHostsContent(podName, containerNam
...
+
+	etcHostContents := `127.0.0.1	localhost
+::1	localhost ip6-localhost ip6-loopback
+fe00::0	ip6-localnet
+ff00::0	ip6-mcastprefix
+ff02::1	ip6-allnodes
+ff02::2	ip6-allrouters`
+
+	// Write the data to a temp file.
+	tmpfile, err := ioutil.TempFile("", "etc-hosts")
```
Comment (liggitt) [inline test/e2e/common/kubelet_etc_hosts.go]: how does writing a random temp file affect the test? this doesn't appear to be used...

Comment (liggitt) [reply test/e2e/common/kubelet_etc_hosts.go]: would we only want to do this in some environments? the referenced issue was specific to the local-up-cluster conformance job, right?

**Comment by liggitt** [comment]
@kubernetes/sig-network-pr-reviews 
adding sig-network for domain expertise

File: test/e2e/common/kubelet_etc_hosts.go (hunk 1 above)
Comment (dims) [reply test/e2e/common/kubelet_etc_hosts.go]: sorry pushed up an update. 

The /etc/hosts file in the environment where pods are launched from is just mounted to the pod. And then there is a check if the file has the string that shows if the file was a managed one or not (using cat on the launched container). So in the CI environment, the /etc/hosts already has the string. so we now create a fresh hosts file and use that.

File: test/e2e/common/kubelet_etc_hosts.go [hunk 2]
```diff
@@ -137,6 +144,27 @@ func (config *KubeletManagedHostConfig) getEtcHostsContent(podName, containerNam
 }
 
 func (config *KubeletManagedHostConfig) createPodSpec(podName string) *v1.Pod {
+
+	etcHostContents := `127.0.0.1	localhost
```
Comment (liggitt) [inline test/e2e/common/kubelet_etc_hosts.go]: seems like creation of this file should be done in setup/cleanup, not inside createPodSpec (which could be called multiple times, or in an unexpected order, in the future)

File: test/e2e/common/kubelet_etc_hosts.go [hunk 3]
```diff
@@ -40,6 +42,7 @@ type KubeletManagedHostConfig struct {
 	hostNetworkPod *v1.Pod
 	pod            *v1.Pod
 	f              *framework.Framework
+	tmpfile        *os.File
```
Comment (liggitt) [inline test/e2e/common/kubelet_etc_hosts.go]: name this better to reflect it's the etcHost file

File: test/e2e/common/kubelet_etc_hosts.go [hunk 4]
```diff
@@ -59,6 +62,10 @@ var _ = framework.KubeDescribe("KubeletManagedEtcHosts", func() {
 
 		By("Running the test")
 		config.verifyEtcHosts()
+
+		if config.tmpfile != nil {
```
Comment (liggitt) [inline test/e2e/common/kubelet_etc_hosts.go]: make a `cleanup()` method and do this in there as a step, just like setup()?

Comment (dims) [reply test/e2e/common/kubelet_etc_hosts.go]: done

File: test/e2e/common/kubelet_etc_hosts.go (hunk 3 above)
Comment (dims) [reply test/e2e/common/kubelet_etc_hosts.go]: done

File: test/e2e/common/kubelet_etc_hosts.go (hunk 2 above)
Comment (dims) [reply test/e2e/common/kubelet_etc_hosts.go]: done

**Comment by dcbw** [comment]
It seems OK to me, but the commit message doesn't really describe what the problem is, or why we care about clean vs. containerized.  What is /etc/hosts missing in the error case that this PR fixes?

**Comment by dims** [comment]
@dcbw please see if the commit message reads better now

**Comment by liggitt** [comment]
isolation LGTM, but would like LGTM from a kubelet test owner

**Comment by dims** [comment]
@derekwaynecarr  @yujuhong  @Random-Liu Can one of you please approve this?

**Comment by m1093782566** [comment]
LGTM

/status approved-for-milestone

**Comment by karataliu** [comment]
One question regarding this change:

I'm running the e2e test from a local dev machine against a remote cluster in cloud.
When running this case, the '/tmp/etc-hosts*' file is created locally on my dev machine but not the vm hosting the pods, thus 'busybox-3' failed to start since the host path '/tmp/etc-hosts*' does not exist.

Seems this case would not support running locally now, any thoughts?

**Comment by dims** [comment]
@karataliu good point, can you please open an issue for me? i think i see the problem, however can you add logs too?

**Pull request merged**